storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -454,76 +460,195 @@ func (s *fileSystemSource) loadShardReadersDataIntoShardResult(
-		}
-
-		s.log.WithFields(
-			xlog.NewField("namespace", nsID.String()),
-			xlog.NewField("shards", len(shardsTimeRanges)),
-		).Infof("filesystem bootstrapper caching block retriever shard indices")
-
-		shards := make([]uint32, 0, len(shardsTimeRanges))
-		for shard := range shardsTimeRanges {
-			shards = append(shards, shard)
-		}
-
-		err = blockRetriever.CacheShardIndices(shards)
+	runResult.RLock()
+	entry, exists = shardResult.AllSeries().Get(id)
+	runResult.RUnlock()
+
+	if exists {
+		// NB(r): In the case the series is already inserted
+		// we can avoid holding onto this ID and use the already
+		// allocated ID.
+		id.Finalize()
+		id = entry.ID
+		tags = entry.Tags
+	} else {
+		tags, err = s.tagsFromTagsIter(tagsIter)
 		if err != nil {
-			return nil, err
+			return fmt.Errorf("unable to decode tags: %v", err)
 		}
 	}
+	tagsIter.Close()
 
-	switch s.opts.ResultOptions().SeriesCachePolicy() {
+	switch seriesCachePolicy {
 	case series.CacheAll:
-		// No checks necessary
+		seg := ts.NewSegment(data, nil, ts.FinalizeHead)
+		seriesBlock.Reset(blockStart, seg)
 	case series.CacheAllMetadata:
-		// Need to check block retriever available
-		if blockRetriever == nil {
-			return nil, fmt.Errorf(
-				"missing block retriever when using series cache metadata for namespace: %s",
-				nsID.String())
+		metadata := block.RetrievableBlockMetadata{
+			ID:       id,
+			Length:   length,
+			Checksum: checksum,
 		}
+		seriesBlock.ResetRetrievable(blockStart, shardRetriever, metadata)
 	default:
-		// Unless we're caching all series (or all series metadata) in memory, we
-		// return just the availability of the files we have
-		bootstrapResult := result.NewDataBootstrapResult()
-		unfulfilled := bootstrapResult.Unfulfilled()
-		for shard, ranges := range shardsTimeRanges {
-			if ranges.IsEmpty() {
-				continue
+		return fmt.Errorf("invalid series cache policy: %s", seriesCachePolicy.String())
+	}
+
+	runResult.Lock()
+	if exists {
+		entry.Blocks.AddBlock(seriesBlock)
+	} else {
+		shardResult.AddBlock(id, tags, seriesBlock)
+	}
+	runResult.Unlock()
+	return nil
+}
+
+func (s *fileSystemSource) readNextEntryAndIndex(
+	r fs.DataFileSetReader,
+	runResult *runResult,
+	segment segment.MutableSegment,
+) error {
+	// If performing index run, then simply read the metadata and add to segment
+	id, tagsIter, _, _, err := r.ReadMetadata()
+	if err != nil {
+		return err
+	}
+
+	// NB(r): Avoiding defer in the hot path here
+	release := func() {
+		// Finalize the ID and tags
+		id.Finalize()
+		tagsIter.Close()
+	}
+
+	idBytes := id.Bytes()
+
+	runResult.RLock()
+	exists, err := segment.ContainsID(idBytes)
+	runResult.RUnlock()
+	if err != nil {
+		release()
+		return err
+	}
+	if exists {
+		release()
+		return nil
+	}
+
+	d, err := convert.FromMetricIter(id, tagsIter)
+	if err != nil {
+		release()
+		return err
+	}
+
+	runResult.Lock()
+	exists, err = segment.ContainsID(idBytes)
+	// ID and tags no longer required below
+	release()
```

Comment:
it looks like FromMetricIter clones the ID so you could probably do:

```golang
d, err := convert.FromMetricIter(id, tagsIter)
release()
if err != nil {
    return err
}

runResult.Lock()
exists, err = segment.ContainsID(d.ID)
```

Reply:
Sure thing.